### PR TITLE
2561 Deactivating inactive user automatically

### LINKED ISSUE #2561: As a sys admin, I need TDP to automatically deactivate accounts that are inactive for 180 days
**Description:**
Per [AC-02(03)](https://github.com/raft-tech/TANF-app/blob/develop/docs/Security-Compliance/Security-Controls/ac-2-3/index.md), TDP must deactivate user accounts after 180 days of inactivity. 

Currently, users are notified of deactivation warnings 10, 3, and 1 day prior to the 180-day deadline.. Sys admins must manually deactivate the user accounts 1 day after the last deactivation warning. To do this: sys admins go into each individual user account that has received their last deactivation warning and remove permissions and change their account approval status to "Deactivated". TDP should take care of this. 

**Acceptance Criteria:** 

- [x] when a user has not logged in for 180 days based on the `last login` date, the permission group should be de-selected for the user. 
- [x] when a user has not logged in for 180 days based on the `last login` date, the account approval status should be changed to `Deactivated`
- [x] the user whose account has been deactivated should receive a deactivate email notification. 
- [x] The sys admin should receive notifications when this happens
- [x] the user should not be able to login to TDP. 
- [x] Run Testing Checklist and confirm all tests pass
- [ ] README is updated, if necessary

**Tasks:**

- [x] update users model as appropriate
- [ ] 
- [ ] tbd
- [ ] Run Testing Checklist and confirm all tests pass

**Supporting Documentation:**
- #1844 
- #2074

### PR DESCRIPTION
**Pull request by raftmsohani**
## Summary of Changes
_Provide a brief summary of changes_
Pull request closes #2561

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d
cd tdrs-backend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d 
```

1. Open http://localhost:3000/ and sign in.
1. Proceed with functional tests as described herein.
1. Test steps should be captured in the demo GIF(s) and/or screenshots below.
> *Demo GIF(s) and screenshots for testing procedure*

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.
+ [x] when a user has not logged in for 180 days based on the last login date, the permission group should be de-selected for the user.
+ [x] when a user has not logged in for 180 days based on the last login date, the account approval status should be changed to Deactivated
+ [x] the user whose account has been deactivated should receive a deactivate email notification.
+ [x] The sys admin should receive notifications when this happens the user should not be able to login to TDP.
+ [x] Run Testing Checklist and confirm all tests pass

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: tdrs-backend/tdpservice/email/tasks.py [hunk 1]
```diff
@@ -16,10 +16,35 @@
...
+def deactivate_users():
+    """Deactivate users that have not logged in in the last 180 days."""
+    users_to_deactivate = User.objects.filter(
+        last_login__lte=datetime.now(tz=timezone.utc) - timedelta(days=180),
+        account_approval_status=AccountApprovalStatusChoices.APPROVED,
+    )
+
+    for user in users_to_deactivate:
+        user.account_approval_status = AccountApprovalStatusChoices.DEACTIVATED
+        user.groups.clear()
```
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/email/tasks.py:34]: Why wouldn't we want to maintain the user's info intact? I don't see it in the ACs.

File: tdrs-backend/tdpservice/scheduling/test/test_user_deactivation_warning.py [hunk 2]
```diff
@@ -30,6 +30,22 @@ def test_deactivation_email_10_days(user, mocker):
...
+    mocker.patch(
+        'tdpservice.email.helpers.account_deactivation_warning.send_deactivation_email',
+        return_value=None
+    )
+    user.groups.add()
+    user.last_login = datetime.now(tz=timezone.utc) - timedelta(days=181)
+    user.account_approval_status = AccountApprovalStatusChoices.APPROVED
+    user.save()
+    tdpservice.email.tasks.deactivate_users()
+    assert user.groups.count() == 0
```
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/scheduling/test/test_user_deactivation_warning.py:45]: related to question above.

File: tdrs-backend/tdpservice/email/tasks.py (hunk 1 above)
Comment (elipe17) [reply tdrs-backend/tdpservice/email/tasks.py:34]: Looks like it is the first AC

File: tdrs-backend/tdpservice/email/tasks.py [hunk 3]
```diff
@@ -16,10 +16,35 @@
 from tdpservice.email.email import automated_email, log
 from tdpservice.email.email_enums import EmailType
 from tdpservice.parsers.util import calendar_to_fiscal
+from tdpservice.email.helpers.account_deactivation_warning import send_deactivation_email
 
 
 logger = logging.getLogger(__name__)
 
+@shared_task
+def deactivate_users():
```
Comment (elipe17) [inline tdrs-backend/tdpservice/email/tasks.py:25]: ACs say that an admin should be notified of the deactivated users. Is that happening somewhere?

File: tdrs-backend/tdpservice/email/tasks.py (hunk 1 above)
Comment (raftmsohani) [reply tdrs-backend/tdpservice/email/tasks.py:34]: the first AC reads: "the permission group should be de-selected for the user"

**Review by andrew-jameson — APPROVED** [review]

**Review by elipe17 — APPROVED** [review]

**Comment by ADPennington** [comment]
@raftmsohani when you get a chance, please add the ACs and testing criteria to this PR.

**Comment by raftmsohani** [comment]
> @raftmsohani when you get a chance, please add the ACs and testing criteria to this PR.

DONE!

**Comment by ADPennington** [comment]
@raftmsohani regarding how to test-- would we need to manually update a user's last login date via django shell?

**Comment by raftmsohani** [comment]
> @raftmsohani regarding how to test-- would we need to manually update a user's last login date via django shell?

That's how I did it with one file submission locally, but if you're testing in Raft/Staging you should have some submitted files

**Comment by raftmsohani** [comment]
> > @raftmsohani regarding how to test-- would we need to manually update a user's last login date via django shell?
>
@ADPennington correct, that's how I did the test

File: tdrs-backend/tdpservice/settings/common.py [hunk 4]
```diff
@@ -475,6 +475,13 @@ class Common(Configuration):
                 'expires': 15.0,
             },
         },
+        'Deactivate Users': {
+            'task': 'tdpservice.email.tasks.deactivate_users',
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/settings/common.py:479]: @raftmsohani what's the name of this periodic task? Is it `Deactivate Users`? I'm not seeing it in the list below, and am unsure how to trigger the task after changing my test user's last login date. I was assuming after setting the last login date to some date > 180 days ago, I would change one of these periodic tasks to trigger the deactivation and email notification...? 

periodic_tasks

Comment (ADPennington) [reply tdrs-backend/tdpservice/settings/common.py:479]: @raftmsohani Im gonna deploy this. just realized i was testing this on another branch 😆

File: tdrs-backend/tdpservice/email/helpers/admin_notifications.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    logger_context = {
+        'user_id': user.id,
+        'object_id': user.id,
+        'object_repr': user.username,
+        'content_type': User,
+    }
+
+    template_path = EmailType.ACCOUNT_DEACTIVATED_ADMIN.value
+    text_message = 'A user account has been deactivated.'
+    subject = 'User Account Deactivated'
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/email/helpers/admin_notifications.py]:
```suggestion
    subject = ' TDP User Account Deactivated due to Inactivity'
```

Comment (ADPennington) [reply tdrs-backend/tdpservice/email/helpers/admin_notifications.py]: recommending this change for clarity @raftmsohani

**Review by ADPennington — CHANGES_REQUESTED** [review]
one small change requested @raftmsohani

**Comment by raftmsohani** [comment]
> one small change requested @raftmsohani

DONE!

**Review by ADPennington — APPROVED** [review]
@raftmsohani LGTM! 🚀  @ttran-hub this PR auto-deactivates user accounts that have been inactive for more than 180 days.  cc: @klinkoberstar 

Please note:

-  the TDP system will check for users who have not logged in for more than 180 days. If it finds a user, it will deactivate their account, send the user a deactivation notification, and send sys admin and ofa admin user groups an email about who was deactivated due to inactivity. One email per deactivated/inactive user will be sent to admins around 9am EST, daily. Artifacts below:

- evidence of auto-deactivation in logentries

logentries_after_auto

- evidence of auto-deactivation in user profiles
_before_
user_before_auto
_after_
user_after_auto

- evidence that user cant sign-in
user_after_autodelete_signin

- evidence of admin email notification of auto deactivation 
admin_email_after auto

**Pull request merged**